generator/standalone.cc

```diff
@@ -0,0 +1,180 @@
+    std::vector<std::string>& filenames,
+    std::string* error_msg) {
+
+  for (const std::string& desc_set : desc_sets) {
+    if (desc_set.empty()) {
+      continue;
+    }
+
+    std::ifstream fstr(desc_set);
+    if (!fstr) {
+      *error_msg = "Could not open file " + desc_set;
+      return false;
+    }
+
+    pb::FileDescriptorSet bin_desc_set;
+    bool parse_result = bin_desc_set.ParseFromIstream(&fstr);
+
+    fstr.close();
+    if (!parse_result) {
+      return false;
+    }
+
+    for (int i = 0; i < bin_desc_set.file_size(); i++) {
+      const auto& fl = bin_desc_set.file(i);
+      auto iter = std::find(packages.begin(), packages.end(), fl.package());
+      if (iter != packages.end()) {
+        filenames.emplace_back(fl.name());
+      }
+    }
+  }
+
+  return true;
+}
+
+void ParseDelimitedArg(const std::string& arg,
+                       std::vector<std::string>& tokens) {
+  // Linux will use ':' and Windows will use ';' as the the delimiter.
+  // Accept both as valid delimiters (not allowed to be used simultaneously).
+  const char* delimiters = ":;";
+  size_t p0 = 0;
+  for (const char* d = delimiters; *d != '\0' && tokens.empty(); d++) {
+    p0 = 0;
+    for (size_t p = arg.find(*d, 0);
+         p != std::string::npos;
+         p0 = p + 1, p = arg.find(*d, p0)) {
+      tokens.emplace_back(arg.substr(p0, p - p0));
+    }
+  }
+  if (arg.size() > p0) {
+    tokens.emplace_back(arg.substr(p0, arg.size() - p0));
+  }
+}
+
+// Converts arguments from what is required by GAPIC generator standalone mode
+// specification (--descriptor, --package, --output) to what is understood by
+// google::protobuf::CommandLineInterface. CommandLineInterface seems like the
+// preferred way of writing GAPIC generator in C++:
+// - it is what is used by protoc main() itself;
+// - it does not spawn a new subprocess if the generator was explicitly
+//   registered by CommandLineInterface::RegisterGenerator(), so it is
+//   performance efficient and good for debugging (runs in same process);
+// - it handles platform-specific file I/O (including writing into zip archive);
+//   this is especially useful for folder operations, since standard library
+//   below C++17 does not have folder I/O abstraction whatsoever.
+//
+bool ConvertCommandLineArgs(int argc,
+                            const char* const argv[],
+                            std::vector<std::string>& args,
+                            std::string* error_msg) {
+  // GAPIC Generator Standalone arguments
+  const std::string desc_arg("--descriptor");
+  const std::string output_arg("--output");
+  const std::string package_arg("--package");
+
+  const std::string desc_set_in_arg("--descriptor_set_in=");
+
+  std::vector<std::string> desc_set_in;
+  std::vector<std::string> packages;
+
+  for (int i = 0; i < argc; i++) {
+    std::string arg(argv[i]);
+    auto pos = arg.find('=', 0);
+    if (pos == std::string::npos) {
+      args.emplace_back(arg);
```

Comment:
Nit:

if (arg.find('=', 0) == std::string::npos) {
  args.emplace_back(std::move(arg));
}

Reply:
This won't work, because `pos` variable is used later in the code (on substring).

Reply:
That makes another strong argument for using abseil StrSplit. The code here is correct, to the best of my reasoning, but doing substring and position manipulation by hand is a red flag for me.